harness/tests/cli/test_auth.py

```diff
@@ -29,74 +34,502 @@
+        #      |    |    |     |    |   token cache state
+        #      |    |    |     |    |    |   Results...
+        #      |    |    |     |    |    |    |
+        #
+        # Explicit user and password, no environment settings, or DET_USER is overridden by
+        # explicitly requested user.
+        Login("y", "y", "ne", "*", "*", "y", CheckCache, UseCache),
+        Login("y", "y", "ne", "*", "*", "n", DoLogin("user", "req_pass"), UseNew),
+        Login("y", "y", "ne", "*", "*", "x", CheckExpired, DoLogin("user", "req_pass"), UseNew),
+        # ---
+        # Explicit user, but password not provided.  Still no (relevant) environment settings.
+        Login("y", "n", "ne", "*", "*", "y", CheckCache, UseCache),
+        Login("y", "n", "ne", "*", "*", "n", DoLogin("user", "prompt_pass"), UseNew),
+        Login("y", "n", "ne", "*", "*", "x", CheckExpired, DoLogin("user", "prompt_pass"), UseNew),
+        # ---
+        # Explicit user and password, DET_USER set but DET_USER_TOKEN not set.  DET_PASS is ignored.
+        Login("y", "y", "y", "*", "n", "y", CheckCache, UseCache),
+        Login("y", "y", "y", "*", "n", "n", DoLogin("user", "req_pass"), UseNew),
+        Login("y", "y", "y", "*", "n", "x", CheckExpired, DoLogin("user", "req_pass"), UseNew),
+        # ---
+        # Explicit user but no password, DET_USER and DET_PASS set, and DET_USER_TOKEN unset.
+        # DET_PASS still ignored since DET_USER/DET_PASS are meant to be processed as a unit, and
+        # an explicitly-requested username overrides that unit.
+        Login("y", "n", "y", "y", "n", "y", CheckCache, UseCache),
+        Login("y", "n", "y", "y", "n", "n", DoLogin("user", "prompt_pass"), UseNew),
+        Login("y", "n", "y", "y", "n", "x", CheckExpired, DoLogin("user", "prompt_pass"), UseNew),
+        # ---
+        # Explicit user but no password, DET_USER set but no other env.  DET_USER is ignored.
+        Login("y", "n", "y", "n", "n", "y", CheckCache, UseCache),
+        Login("y", "n", "y", "n", "n", "n", DoLogin("user", "prompt_pass"), UseNew),
+        Login("y", "n", "y", "n", "n", "x", CheckExpired, DoLogin("user", "prompt_pass"), UseNew),
+        # ---
+        # DET_USER_TOKEN is overridden by a configured cache (the on-cluster `det user login` case).
+        Login("*", "*", "ye", "*", "y", "y", CheckCache, UseCache),
+        # DET_USER_TOKEN can be used if user is explicit, so long as DET_USER matches explicit user.
+        Login("y", "*", "y", "*", "y", "n", Use("user", "env_token")),
+        # No explicit user; token in env is used if cache is missing or invalid.
+        Login("n", "*", "y", "*", "y", "n", Use("user", "env_token")),
+        Login("n", "*", "y", "*", "y", "x", CheckExpired, Use("user,", "env_token")),
+        # No explicit user; token in env is overridden by a configured cache.
+        # (the on-cluster `det user login` case).
+        Login("n", "*", "ye", "*", "y", "y", CheckCache, UseCache),
```

Comment:
Again (and I think for my next few comments following) I think we can check for returned user if we remove `"e"` from `DET_USER` arg here.